### PR TITLE
Fix Configuration with IList and ICollection

### PR DESCRIPTION
**Pull request by tarekgh**
https://github.com/dotnet/runtime/issues/77725

In .NET 7.0 we had some code refactoring and fixes in configuration. This change had been handling the binding to `ICollction<T>` and `IList<T>` by creating underlying `Array<T>` object. It is obvious the returned object is not expandable which means users cannot add more items to that object. This is wrong and regression from previous versions of .NET/Core. The fix here is in case of binding to `ICollction<T>` and `IList<T>`, we'll create `List<T>` object instead which is expandable and allow adding more items to such object. 

#77725 is a good example demonstrating this problem.

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @SteveDunn @ericstj

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/pull/68133
https://github.com/dotnet/runtime/pull/73267

**Review by SteveDunn — APPROVED** [review]
LGTM

**Review by maryamariyan — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 1]
```diff
@@ -315,6 +315,19 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
                     }
                 }
 
+                Type? iCollectionInterfaceType = GetICollectionInterfaceType(type);
+                if (iCollectionInterfaceType is not null)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:319]: This feels like it is duplicated with

https://github.com/dotnet/runtime/blob/9df6ea21007319967975dc9985413bb6518287da/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L371-L374

I think these should be combined.

I wonder if we can just delete this code here, and

1. Let the linked code handle the case when an object already exists.
2. Add back in the "make a new collection instance" code inside the `if (bindingPoint.Value is null)` check. See 
https://github.com/dotnet/runtime/blob/15d285f325f5ad209cd2ede3262e5f65ccf1bd73/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L367-L374

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs [hunk 2]
```diff
@@ -393,6 +393,11 @@ public void AlreadyInitializedListInterfaceBinding()
             Assert.Equal("val1", list[2]);
             Assert.Equal("val2", list[3]);
             Assert.Equal("valx", list[4]);
+
+            // Ensure expandability of the returned list
+            options.AlreadyInitializedListInterface.Add("ExtraItem");
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs:398]: Do we have a test for an `IList` or `ICollection` property, where the backing field hasn't been set yet?  Basically a test for the `if (bindingPoint.Value is null)` check.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 3]
```diff
@@ -851,8 +885,6 @@ private static bool IsArrayCompatibleInterface(Type type)
 
             Type genericTypeDefinition = type.GetGenericTypeDefinition();
             return genericTypeDefinition == typeof(IEnumerable<>)
-                || genericTypeDefinition == typeof(ICollection<>)
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:854]: I am trying to reduce the change here as I am planning to service this fix to 7.0. We can have other follow-up changes for 8.0 as needed. Let me know if this reasonable to you.

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:854]: The code should still be reasonable. With these changes, we now have duplicate checks here:

https://github.com/dotnet/runtime/blob/9df6ea21007319967975dc9985413bb6518287da/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L303-L312

The first and the last `if` conditions are now identical.

[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:854]: Looking more, I see your point. I'll try to do your suggestion.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationCollectionBindingTests.cs:398]: This test case already using `UninitializedCollectionsOptions` which include the uninitialized `ICollection`.  The extra added code to this test code is testing the returned collection to allow adding more items to it. 
The next test case `CanBindUninitializedIList` is test `IList` and I did the same to ensure the returned list allows adding more items.

**Comment by tarekgh** [comment]
https://github.com/dotnet/runtime/pull/77582

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 4]
```diff
@@ -350,12 +346,24 @@ private static void BindProperty(PropertyInfo property, object instance, IConfig
...
+                    if (interfaceGenericType is not null &&
+                        (interfaceGenericType == typeof(ICollection<>) || interfaceGenericType == typeof(IList<>)))
+                    {
+                        // For ICollection<T> and IList<T> we bind them to mutable List<T> type.
+                        Type genericType = typeof(List<>).MakeGenericType(type.GenericTypeArguments[0]);
+                        bindingPoint.SetValue(Activator.CreateInstance(genericType));
+                    }
+                    else
+                    {
+                        // For other mutable interfaces like ICollection<> and ISet<>, we prefer copying values and setting them
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:360]: I don't think this comment makes sense any more.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 5]
```diff
@@ -584,9 +592,10 @@ private static bool CanBindToTheseConstructorParameters(ParameterInfo[] construc
                 return;
             }
 
-            MethodInfo tryGetValue = dictionaryType.GetMethod("TryGetValue", BindingFlags.Public | BindingFlags.Instance)!;
-
             Debug.Assert(dictionary is not null);
+
+            MethodInfo tryGetValue = dictionary.GetType().GetMethod("TryGetValue", BindingFlags.Public | BindingFlags.Instance)!;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:597]: (nit) can pull out `dictionary.GetType()` into a variable and reuse it for both calls.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for fixing this @tarekgh!

**Pull request merged**